### PR TITLE
support pp accuracy calculation

### PR DESCRIPTION
**Pull request by wtmlon**

### CODE REVIEW

File: paddlenlp/trainer/trainer.py [hunk 1]
```diff
@@ -3276,9 +3295,42 @@ def prediction_pipeline_step(
...
+        loss, _, labels = self.prediction_pipeline_step(*args, **kwargs)
+        logits = None
+        # infohub
+        hcg = fleet.get_hybrid_communicate_group()
+        if hcg:
+            pp_group = hcg.get_pipe_parallel_group()
+            if pp_group.nranks > 1:
+                logit_shape = [[]]
+                if "pp_logits" in infohub:
+                    logits = paddle.concat(infohub["pp_logits"], axis=0)
```
Comment (DesmonDay) [inline paddlenlp/trainer/trainer.py]: 这里为啥是concat了？不是很理解

File: paddlenlp/trainer/trainer.py [hunk 2]
```diff
@@ -3268,6 +3276,17 @@ def prediction_pipeline_step(
                 labels = None
             inputs = inputs.pop("input_ids")
 
+        # evaluation dont support drop last,
+        # so set the `accumulate_steps` to actually
+        # eval batch size.
+        model_config_backup = model.accumulate_steps
```
Comment (wawltor) [inline paddlenlp/trainer/trainer.py]: 这个命名是不是不太规范？ 很明显这个又不是一个model config

File: paddlenlp/trainer/trainer.py [hunk 3]
```diff
@@ -3276,9 +3295,25 @@ def prediction_pipeline_step(
...
+    def prediction_pipeline_step_with_logits_acc(
+        self,
+        *args,
+        **kwargs,
+    ):
+        loss, _, labels = self.prediction_pipeline_step(*args, **kwargs)
+        logits = None
+        if "pp_logits" in infohub:
+            logits = paddle.concat(infohub["pp_logits"], axis=0)
+            logits = logits._copy_to(paddle.framework._current_expected_place(), False)
```
Comment (wawltor) [inline paddlenlp/trainer/trainer.py]: 这里拷贝的原因是pp_logits是放在cpu memory 或者 cuda pin memory？

Comment (wtmlon) [reply paddlenlp/trainer/trainer.py]: 是的，如果这里不放在cpu或者 pin memory 会在 concat 的时候造成增加两倍 logits 大小的峰值显存，导致 OOM

File: paddlenlp/trainer/trainer.py [hunk 4]
```diff
@@ -3312,6 +3347,8 @@ def prediction_step(
         if self.args.pipeline_parallel_degree > 1:
             # hack for pipeline mode
             inputs = self._prepare_inputs(inputs)
+            if self.args.metric_for_best_model == "accuracy":
```
Comment (lugimzzz) [inline paddlenlp/trainer/trainer.py]: 这个建议不要放在trainer，放在SFTTrainer更加合理

File: paddlenlp/trainer/trainer.py (hunk 2 above)
Comment (lugimzzz) [reply paddlenlp/trainer/trainer.py]: 不能只修改accumulate_steps而不修改micro_batch_size，
https://github.com/PaddlePaddle/Paddle/blob/52f55e159fd8235c841985578e380b9d9dc3a220/python/paddle/distributed/fleet/meta_parallel/pipeline_parallel.py#L271C1-L275C31

File: paddlenlp/trainer/trainer.py [hunk 5]
```diff
@@ -3268,6 +3276,17 @@ def prediction_pipeline_step(
...
+        # evaluation dont support drop last,
+        # so set the `accumulate_steps` to actually
+        # eval batch size.
+        model_config_backup = model.accumulate_steps
+        if isinstance(inputs, tuple):
+            input_ids = inputs[0]
+        else:
+            input_ids = inputs
+
+        model.accumulate_steps = input_ids.shape[0]
```
Comment (lugimzzz) [inline paddlenlp/trainer/trainer.py]: 要不就要把model.micro_batch_size直接设为1

File: tests/trainer/test_unified_checkpoint.py [hunk 6]
```diff
@@ -81,6 +81,7 @@
     "fp16_opt_level": "O2",
     "max_grad_norm": 1.0,
     "dataloader_num_workers": 0,
+    "metric_for_best_model": "accuracy",
```
Comment (lugimzzz) [inline tests/trainer/test_unified_checkpoint.py]: 后续也在开源模型适配

**Review by lugimzzz — DISMISSED** [review]
LGTM

**Review by lugimzzz — APPROVED** [review]
lgtm

**Review by wawltor — APPROVED** [review]
LGTM
